src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs

```diff
@@ -61,24 +61,42 @@ protected override void EndProcess ()
 			throw new InvalidOperationException ($"Unable to find {TypeName} type");
 		}
 
-		// Java -> .NET mapping
-		var orderedJavaToDotnetMapping = TypeMappings.OrderBy (kvp => Hash (kvp.Key)).ToArray ();
-		var javaClassNameHashes = orderedJavaToDotnetMapping.Select (kvp => Hash (kvp.Key)).ToArray ();
-		var types = orderedJavaToDotnetMapping.Select (kvp => SelectTypeDefinition (kvp.Key, kvp.Value));
+		var typeMappingRecords = TypeMappings
+			.Select (kvp =>
+				new TypeMapRecord {
+					JavaClassName = kvp.Key,
+					Types = kvp.Value.ToArray (),
+					Context = Context,
+				})
+			.ToArray ();
 
-		GenerateHashes (javaClassNameHashes, methodName: "get_JavaClassNameHashes");
-		GenerateGetTypeByIndex (types);
-		GenerateStringSwitchMethod (type, "GetJavaClassNameByTypeIndex", orderedJavaToDotnetMapping.Select (kvp => kvp.Key).ToArray ());
+		// Java -> .NET mapping
+		{
+			var orderedJavaToDotnetMapping = typeMappingRecords.OrderBy (record => Hash (record.JavaClassName)).ToArray ();
+			var javaClassNames = orderedJavaToDotnetMapping.Select (record => record.JavaClassName).ToArray ();
+			var javaClassNameHashes = javaClassNames.Select (Hash).ToArray ();
+			var types = orderedJavaToDotnetMapping.Select (record => record.SelectTypeDefinition ());
+
+			GenerateHashes (javaClassNameHashes, methodName: "get_JavaClassNameHashes");
+			GenerateGetTypeByIndex (types);
+			GenerateStringSwitchMethod (type, "GetJavaClassNameByTypeIndex", javaClassNames);
+		}
 
 		// .NET -> Java mapping
-		var orderedManagedToJavaMapping = TypeMappings.SelectMany(kvp => kvp.Value.Select (type => new KeyValuePair<string, string>(kvp.Key, GetAssemblyQualifiedTypeName (type)))).OrderBy (kvp => Hash (kvp.Value)).ToArray ();
-		var assemblyQualifiedTypeNames = orderedManagedToJavaMapping.Select (kvp => kvp.Value).ToArray ();
-		var assemblyQualifiedTypeNameHashes = assemblyQualifiedTypeNames.Select (Hash).ToArray ();
-		var javaClassNames = orderedManagedToJavaMapping.Select (kvp => kvp.Key).ToArray ();
-
-		GenerateHashes (assemblyQualifiedTypeNameHashes, methodName: "get_TypeNameHashes");
-		GenerateStringSwitchMethod (type, "GetJavaClassNameByIndex", javaClassNames);
-		GenerateStringSwitchMethod (type, "GetAssemblyQualifiedTypeNameByJavaClassNameIndex", assemblyQualifiedTypeNames);
+		{
+			var orderedManagedToJavaMapping = typeMappingRecords
+				.SelectMany (record => record.Flatten ())
+				.OrderBy (record => Hash (record.AssemblyQualifiedTypeName))
+				.ToArray ();
+
+			var assemblyQualifiedTypeNames = orderedManagedToJavaMapping.Select (record => record.AssemblyQualifiedTypeName).ToArray ();
```

Comment:
@grendello are we till trying to avoid using Linq in the tasks and linker steps?

Reply:
I'd hope so, it's still slower than plain, old, boring, hand-written code :)

Reply:
I can remove the LINQ queries in a follow-up PR.